nav2_behavior_tree/test/plugins/decorator/distance_controller_test.cpp

```diff
@@ -0,0 +1,177 @@
+  }
+
+  void SetUp()
+  {
+    node_ = new nav2_behavior_tree::DistanceController("distance_controller", *config_);
+    node_->setChild(dummy_node_);
+  }
+
+  void TearDown()
+  {
+    node_ = nullptr;
+  }
+
+protected:
+  static nav2_behavior_tree::TransformHandler * transform_handler_;
+  static BT::NodeConfiguration * config_;
+  static nav2_behavior_tree::DistanceController * node_;
+  static nav2_behavior_tree::DummyNode * dummy_node_;
+};
+
+nav2_behavior_tree::TransformHandler * DistanceControllerTestFixture::transform_handler_ = nullptr;
+BT::NodeConfiguration * DistanceControllerTestFixture::config_ = nullptr;
+nav2_behavior_tree::DistanceController * DistanceControllerTestFixture::node_ = nullptr;
+nav2_behavior_tree::DummyNode * DistanceControllerTestFixture::dummy_node_ = nullptr;
+
+TEST_F(DistanceControllerTestFixture, test_initial_status_is_idle)
+{
+  EXPECT_EQ(node_->status(), BT::NodeStatus::IDLE);
+}
+
+TEST_F(DistanceControllerTestFixture, test_failure_when_child_is_idle)
+{
+  dummy_node_->setStatus(BT::NodeStatus::IDLE);
+  EXPECT_EQ(node_->executeTick(), BT::NodeStatus::FAILURE);
+}
+
+TEST_F(DistanceControllerTestFixture, test_failure_when_child_fails)
+{
+  dummy_node_->setStatus(BT::NodeStatus::FAILURE);
+  EXPECT_EQ(node_->executeTick(), BT::NodeStatus::FAILURE);
+}
+
+TEST_F(DistanceControllerTestFixture, test_running_when_child_is_running)
+{
+  dummy_node_->setStatus(BT::NodeStatus::RUNNING);
+  EXPECT_EQ(node_->executeTick(), BT::NodeStatus::RUNNING);
+}
+
+TEST_F(DistanceControllerTestFixture, test_behavior)
+{
+  EXPECT_EQ(node_->status(), BT::NodeStatus::IDLE);
+
+  dummy_node_->setStatus(BT::NodeStatus::SUCCESS);
+  EXPECT_EQ(node_->executeTick(), BT::NodeStatus::SUCCESS);
+  EXPECT_EQ(dummy_node_->status(), BT::NodeStatus::IDLE);
+
+  geometry_msgs::msg::PoseStamped pose;
+  pose.pose.position.x = 0.5;
+  pose.pose.position.y = 0;
+  pose.pose.position.z = 0;
+  pose.pose.orientation.x = 0;
+  pose.pose.orientation.y = 0;
+  pose.pose.orientation.z = 0;
+  pose.pose.orientation.w = 1;
+
+  double traveled = 0;
+
+  transform_handler_->updateRobotPose(pose.pose);
+
+  // Wait for transforms to actually update
+  while (traveled < 0.1) {
+    if (nav2_util::getCurrentPose(pose, *transform_handler_->getBuffer())) {
+      traveled = std::sqrt(
+        pose.pose.position.x * pose.pose.position.x + pose.pose.position.y * pose.pose.position.y);
+    }
+    EXPECT_EQ(node_->executeTick(), BT::NodeStatus::RUNNING);
+  }
+
+  for (int i = 1; i < 10; i++) {
+    pose.pose.position.x = i * 1.1;
+    transform_handler_->updateRobotPose(pose.pose);
+
+    // Wait for transforms to actually update
+    while (traveled < i) {
+      if (nav2_util::getCurrentPose(pose, *transform_handler_->getBuffer())) {
+        traveled = std::sqrt(
+          pose.pose.position.x * pose.pose.position.x +
+          pose.pose.position.y * pose.pose.position.y);
+      }
+    }
+
+    dummy_node_->setStatus(BT::NodeStatus::SUCCESS);
```

Comment:
Why do you need to set the status to success first?

Reply:
This line sets the status of the dummy child node to SUCCESS, to basically mimic an action node that this decorator controls. After an update of N meters, DistanceController sets its child back to IDLE. This logic is internally handled by the tick engine.